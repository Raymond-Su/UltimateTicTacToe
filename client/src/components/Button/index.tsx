import React, { FC, ReactNode } from 'react';
import classNames from 'classnames';
import './Button.scss';

interface ButtonProps {
  id?: string;
  primary?: boolean;
  onClick?: () => void;
  children?: ReactNode;
}

const Button: FC<ButtonProps> = ({
  id,
  primary = false,
  onClick,
  children
}: ButtonProps) => {
  return (
    <button
      type="button"
      id={id}
      className={classNames('btn', { 'btn-primary ': primary })}
      onClick={onClick}
    >
      {children}
    </button>
  );
};

export default Button;
