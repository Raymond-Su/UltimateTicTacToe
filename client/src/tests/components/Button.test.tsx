import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';

import Button from '../../components/Button';

const mockFunction = jest.fn();

describe('<Button />', () => {
  test('Renders component', () => {
    const { getByTestId } = render(<Button />);
    const button = getByTestId('btn');
    expect(button).toBeInstanceOf(HTMLButtonElement);
    expect(button).toHaveClass('btn');
    expect(button).not.toHaveClass('btn-primary');
  });

  test('styled correctly when primary property is applied', () => {
    const { getByTestId } = render(<Button primary />);
    const button = getByTestId('btn');
    expect(button).toHaveClass('btn', 'btn-primary');
  });

  test('btn to have id when id property is applied', () => {
    const { getByTestId } = render(<Button id="test-id" />);
    const button = getByTestId('btn');
    expect(button).toHaveAttribute('id', 'test-id');
  });

  test('renders children', () => {
    const { getByTestId } = render(
      <Button>
        <p data-testid="children">Children</p>
      </Button>
    );
    const button = getByTestId('btn');
    const children = getByTestId('children');
    expect(button).toContainElement(children);
  });

  test('handles function call when onClick property is added', () => {
    const { getByTestId } = render(<Button onClick={mockFunction} />);
    const button = getByTestId('btn');
    fireEvent.click(button);
    expect(mockFunction).toBeCalledTimes(1);
  });

  test('Matches the snapshot', () => {
    const button = create(
      <Button onClick={mockFunction} primary id="snapshot-btn">
        Snapshot Button
      </Button>
    );
    expect(button.toJSON()).toMatchSnapshot();
  });
});
