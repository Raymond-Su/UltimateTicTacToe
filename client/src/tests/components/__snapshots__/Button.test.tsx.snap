// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Matches the snapshot 1`] = `
<button
  className="btn btn-primary "
  data-testid="btn"
  id="snapshot-btn"
  onClick={[MockFunction]}
  type="button"
>
  Snapshot Button
</button>
`;
